xmodule/tasks.py

```diff
@@ -0,0 +1,316 @@
+"""
+This module contains tasks for asynchronous execution of library xblocks.
+"""
+import hashlib
+
+from celery import shared_task
+from celery.utils.log import get_task_logger
+from django.contrib.auth.models import User  # lint-amnesty, pylint: disable=imported-auth-user
+from django.core.exceptions import PermissionDenied
+from opaque_keys.edx.keys import UsageKey
+from opaque_keys.edx.locator import (
+    BlockUsageLocator,
+    LibraryLocatorV2,
+    LibraryUsageLocator,
+    LibraryUsageLocatorV2
+)
+from search.search_engine_base import SearchEngine
+from user_tasks.models import UserTaskStatus
+from user_tasks.tasks import UserTask
+from xblock.fields import Scope
+
+from common.djangoapps.student.auth import has_studio_write_access
+from openedx.core.djangoapps.content_libraries import api as library_api
+from openedx.core.djangoapps.content_libraries.models import ContentLibrary
+from openedx.core.djangoapps.xblock.api import load_block
+from openedx.core.lib import blockstore_api
+from xmodule.capa_module import ProblemBlock
+from xmodule.library_content_module import ANY_CAPA_TYPE_VALUE
+from xmodule.modulestore import ModuleStoreEnum
+from xmodule.modulestore.django import modulestore
+from xmodule.modulestore.exceptions import ItemNotFoundError
+
+LOGGER = get_task_logger(__name__)
+
+
+def normalize_key_for_search(library_key):
+    """ Normalizes library key for use with search indexing """
+    return library_key.replace(version_guid=None, branch=None)
+
+
+def _import_block(store, user_id, source_block, dest_parent_key):
+    """
+    Recursively import a blockstore block and its children. See import_from_blockstore above.
+    """
+    def generate_block_key(source_key, dest_parent_key):
```

Comment:
Can this be separate function?

Reply:
i just took it from https://github.com/raccoongang/edx-platform/pull/2440/files#diff-07ba8e5cc4e46eefe5c9a3034b05d58f90001587ccd63fd2910ada28d4b22bafL295
I get an error message when I take this function out. I think we can leave it as it was.